components/push/src/communications.rs

```diff
@@ -200,9 +202,7 @@ impl Connection for ConnectHttp {
         {
             Ok(v) => v,
             Err(e) => {
-                return Err(
-                    CommunicationServerError(format!("Could not fetch endpoint: {}", e)).into(),
-                );
+                return Err(EndpointError(format!("Could not fetch endpoint: {}", e)).into());
```

Comment:
So IIUC, this will throw `EndpointError` is there is e.g. a network-level error while subscribing, but will continue throwing a `CommunicationError` if the server successfully returns a 400-level HTTP error code.

From the PR description it sounds like you may also want to return `EndpointError` in the case of an error response from the server, is that accurate? If so, we'll also need to update the places below where we `return Err(...)`.

Reply:
That depends, from what I understand, the other errors are related to the client-server communication which could be a general error and not for the specific case where we can successfully establish an HTTP connection but the retrieving the endpoint fails.

`is_server_error` seems to relate to 500/600 http error codes, and `is_client_error` is for 400/500.

This line with `CommunicationServerError(format!("Could not parse response: {:?}", e))` may be a candidate, but I'm not sure what "Could not parse response" refers to. Is it a parsing error from the corrupted JSON on the server or from a network failure?

Reply:
> I'm not sure what "Could not parse response" refers to. Is it a parsing error from the corrupted JSON on the
> server or from a network failure?

This is if the server returns invalid JSON in its otherwise-successful response.